electronics_lib/PowerConditioning.py

```diff
@@ -263,3 +264,116 @@ def connected_from(self, gnd: Optional[Port[VoltageLink]] = None, pwr_hi: Option
     if pwr_lo is not None:
       cast(Block, builder.get_enclosing_block()).connect(pwr_lo, self.pwr_lo)
     return self
+
+
+class PmosReverseProtection(PowerConditioner, KiCadSchematicBlock, Block):
+  """Reverse polarity protection using a PMOS. This method has lower power loss over diode-based protection.
+  100R-330R is good but 1k-50k can be used for continuous load.
+  Ref: https://components101.com/articles/design-guide-pmos-mosfet-for-reverse-voltage-polarity-protection
+  """
+
+  @init_in_parent
+  def __init__(self, gate_resistor: RangeLike = 10 * kOhm(tol=0.05), rds_on: RangeLike = (0, 0.1) * Ohm):
+    super().__init__()
+    self.gnd = self.Port(Ground.empty(), [Common])
+    self.pwr_in = self.Port(VoltageSink.empty())
+    self.pwr_out = self.Port(VoltageSource.empty())
+
+    self.gate_resistor = self.ArgParameter(gate_resistor)
+    self.rds_on = self.ArgParameter(rds_on)
+
+  def contents(self):
+    super().contents()
+    output_current_draw = self.pwr_out.link().current_drawn
+    self.fet = self.Block(Fet.PFet(
+      drain_voltage=(0, self.pwr_out.link().voltage.upper()),
+      drain_current=output_current_draw,
+      gate_voltage=(- self.pwr_out.link().voltage.upper(), self.pwr_out.link().voltage.upper()),
+      rds_on=self.rds_on,
+    ))
+
+    self.res = self.Block(Resistor(self.gate_resistor))
+    # TODO: generate zener diode for high voltage applications
+    #  self.diode = self.Block(ZenerDiode(self.clamp_voltage))
+
+    self.import_kicad(
+      self.file_path("resources", f"{self.__class__.__name__}.kicad_sch"),
+      conversions={
+        'pwr_in': VoltageSink(
+          current_draw=output_current_draw,
+        ),
+        'pwr_out': VoltageSource(
+          voltage_out=self.pwr_in.link().voltage,
+        ),
+        'gnd': Ground(),
+      })
+
+
+class PmosChargerReverseProtection(PowerConditioner, KiCadSchematicBlock, Block):
+  """Charging capable a battery reverse protection using PMOS transistors. The highest battery voltage is bounded by the
+  transistors' Vgs/Vds. There is also a rare case when this circuit being disconnected when a charger is connected first.
+  But always reverse protect. R1 and R2 are the pullup bias resistors for mp1 and mp2 PFet.
+  More info at: https://www.edn.com/reverse-voltage-protection-for-battery-chargers/
+  """
+
+  @init_in_parent
+  def __init__(self, r1_val: RangeLike = 100 * kOhm(tol=0.01), r2_val: RangeLike = 100 * kOhm(tol=0.01),
```

Comment:
```suggestion
  def __init__(self, r1_val: RangeLike = 10 * kOhm(tol=0.01), r2_val: RangeLike = 10 * kOhm(tol=0.01),
```

How come R1 and R2 default to 100k when the comments say the typical range is 1k-50k? Maybe set it to 10k to make it consistent with the non-charging-capable PMOS circuit and also within the typical range?

Reply:
The comments are for PmosReverseProtection and for the PmosChargerReverseProtection, the R1, R2 are 100k in the reference